0xProject/0x-monorepo
# Overflow free getPartialAmount and isRoundingError

## Description

Current implementation of `getPartialAmount` and `isRoundingError` can overflow, causing a failure of the transaction. These overflows do not return revert reasons and we never clearly specified when they could occur. This PR implements math that gives the correct result in all cases, without overflow.

This PR also adds the requirement that `numerator / denominator <= 1`. This guarantees that a result always exists (without it the result could be more than 2^256). In several places the order of the arguments had `target` and `numerator` swapped, this is corrected.

In `marketBuyOrder` and `marketBuyOrderNoThrow` the value acting as `numerator` could become larger than `denominator`. This creates a potential for an unsolvable overflow. It is fixed by taking the minimum of the outstanding amount and the order amount.

The math in `getPartialAmount` uses a whole collection of theorems I developed some time ago. They are derived and explained in a series of Medium posts:

1. https://medium.com/wicketh/mathemagic-chinese-remainder-theorem-a5604dfd96b0
2. https://medium.com/wicketh/mathemagic-full-multiply-27650fec525d
3. https://medium.com/wicketh/mathemagic-512-bit-division-in-solidity-afa55870a65
4. https://medium.com/wicketh/mathemagic-multiplicative-inverses-b093e2a0a6d3

### Error handling

This PR introduces a new revert reason `NUMERATOR_GT_DENOMINATOR` when `getPartialAmount` is called with a fraction larger than one. This should never happen under correct operation of the exchange. (It would indicate an order that is more than 100% filled).

### Gas cost and optimization

Impact on gas consumption has not been measured yet. It should be good for `getPartialAmount` since it only adds significant work when the current implementation would have failed. `isRoundingError` became substantially more complex, but it is still a rather insignificant function gas wise.

The `isRoundingError` would benefit significantly from being merged with `getPartialAmount` as they share a lot of intermediate values. We could have a `getPartialAmount` and `getPartialAmountWithBoundedError` where the latter would revert with `EXCESSIVE_ROUNDING_ERROR` when the error exceeds the treshold.


## Testing instructions

* Check the Math.
* Check the Solidity implementation.
* Check the JS reference implementation (requires only math, not Math).
* Uncomment and/or add more test values.
* Run the combinatorial tests to check equivalence of Solidity and reference.
* Hand compute the result on select values.
* Check changes to the Exchange contracts.
* Especially check changes to `marketBuyOrder` and `marketBuyOrderNoThrow`.

## Types of changes




## Checklist:
